### PR TITLE
Implement BlockValueNet and associated training functions

### PR DESCRIPTION
**Pull request by alanlujan91**
This pull request introduces several enhancements to the `skagent` library, focusing on implementing Bellman residual computation, creating a value network for dynamic programming problems, and enabling joint training of policy and value networks. It also includes updates to the testing framework to support these new features.

### Bellman Residual and Loss Functions:
* Added `estimate_bellman_residual` and `get_bellman_equation_loss` functions in `src/skagent/algos/maliar.py` to compute and utilize Bellman equation residuals for dynamic programming problems. These functions enable accurate assessment of value functions and integrate seamlessly with the Maliar method.

### Value Network Implementation:
* Introduced the `BlockValueNet` class in `src/skagent/ann.py`, designed to approximate value functions using neural networks. This class includes methods for computing value estimates and retrieving callable value functions for loss evaluation.

### Joint Training of Networks:
* Added `train_block_value_nn` and `train_block_value_and_policy_nn` functions in `src/skagent/ann.py` to train value and policy networks individually or jointly. These functions utilize loss functions and optimizers to iteratively improve network performance.

### Testing Enhancements:
* Updated `tests/test_ann.py` to include new imports (`skagent.model` and `HARK.distributions`) to support testing of the added value network and Bellman residual functionality.

### CODE REVIEW

File: src/skagent/algos/maliar.py [hunk 1]
```diff
@@ -308,4 +308,183 @@ def maliar_training_loop(
...
+    rf = create_reward_function(block, agent)
+
+    # Compute immediate reward
+    immediate_reward = rf(states_t, shocks_t, controls_t, parameters)[reward_sym]
+
+    # Compute next states
+    next_states = tf(states_t, shocks_t, controls_t, parameters)
+
+    # Compute continuation value using value network
+    continuation_values = value_network(next_states, shocks_t, parameters)
```
Comment (sbenthall) [inline src/skagent/algos/maliar.py:387]: It looks to me like the `shocks_t` input is doing the work of the shocks in period `t` (in the calculation of the 'immediate reward') as well `t+1` in the computation of the continuation value (when passed to the value network).

I'm surprised by this. Do we actually need two different samples of the shocks here, so we have independent draws for the `t` and `t+1` shocks at work?

Comment (sbenthall) [reply src/skagent/algos/maliar.py:387]: Since automated tests were passing despite this being a bug, there should be an automated test (maybe from one of the benchmarks) that tests for for the solver's performance on a problem with genuine time-dependence.

**Comment by sbenthall** [comment]
This looks good! Is it ready to merge from your perspective @alanlujan91 ?

**Comment by sbenthall** [comment]
Thanks, @alanlujan91 ! Merged.
